src/x86/sse41.rs

```diff
@@ -238,11 +239,42 @@ pub unsafe fn _mm_max_epi32(a: i32x4, b: i32x4) -> i32x4 {
 /// maximum values.
 #[inline(always)]
 #[target_feature = "+sse4.1"]
-#[cfg_attr(test, assert_instr(pmaxud, imm8 = 0))]
+#[cfg_attr(test, assert_instr(pmaxud))]
 pub unsafe fn _mm_max_epu32(a: u32x4, b: u32x4) -> u32x4 {
     pmaxud(a, b)
 }
 
+/// Convert packed 32-bit integers from `a` and `b` to packed 16-bit integers using unsigned saturation
+#[inline(always)]
+#[target_feature = "+sse4.1"]
+#[cfg_attr(test, assert_instr(packusdw))]
+pub unsafe fn _mm_packus_epi32(a: i32x4, b: i32x4) -> u16x8 {
+    packusdw(a, b)
+}
+
+/// Compare packed 64-bit integers in `a` and `b` for equality
+#[inline(always)]
+#[target_feature = "+sse4.1"]
+#[cfg_attr(test, assert_instr(pcmpeqq))]
+pub unsafe fn _mm_cmpeq_epi64(a: i64x2, b: i64x2) -> i64x2 {
+    a.eq(b)
+}
+
+/// Sign extend packed 8-bit integers in `a` to packed 16-bit integers
+#[target_feature = "+sse4.1"]
+#[cfg_attr(test, assert_instr(pmovsxbw))]
+pub unsafe fn _mm_cvtepi8_epi16(a: i8x16) -> i16x8 {
+    simd_cast::<::v64::i8x8, _>(simd_shuffle8(a, a, [0, 1, 2, 3, 4, 5, 6, 7]))
+}
+
+/// Sign extend packed 8-bit integers in `a` to packed 32-bit integers
+#[target_feature = "+sse4.1"]
+#[cfg_attr(test, assert_instr(pmovsxbd))]
+pub unsafe fn _mm_cvtepi8_epi32(a: i8x16) -> i32x4 {
+    let cast = simd_cast::<_, ::v512::i32x16>(a);
```

Comment:
Why are you using a 512 bit vector here? What does Clang do?

Reply:
I can't do this:
```rust
simd_cast::<::v32::i8x4, _>(simd_shuffle4(a, a, [0, 1, 2, 3]))
```
Because there is no `i8x4` type (note that `v32` is an hypothetical module containing the type `i8x4`)

So I instead convert first `i8x16` to `i32x16` and then get the `i32x4` with `simd_suffle4`, instead of the other way around. 

I do realize that this might no be a good way to solve this even do it generated the right instruction. Will the usage of a `v512` type require 512 bit wide registers in a machine that might not have them?

Reply:
Can you do what Clang does? It doesn't use an `i8x4` or a 512 bit vector type: https://github.com/llvm-mirror/clang/blob/a38ba9770a70056f9fdd6c71f819e5db45a105e4/lib/Headers/smmintrin.h#L1283-L1289

Reply:
What clang does would be this line:
```rust
simd_cast::<::v32::i8x4, _>(simd_shuffle4(a, a, [0, 1, 2, 3]))
```
But rust's `simd_suffle4` has to return a `i8x4` which does not exist

Reply:
@p32blo Hmm sorry, I guess I'm not seeing it. Can you point out where the 32 bit vector is being used? I see `__v4si` and `__v16qs`, both of which I think are 128 bit vectors?

Reply:
Ok, we have a 128 bit vector with 8 bit elements, so there are 16 elements. First, with `simd_shuffle` we get the first 4 elements. This leaves us with 32 bits. Then with `simd_cast`, we sign extend this 4 elements to be 32 bits and this brings us back to the 128 bit total.

The problem were is when we take the first 4 elements we need a 32 bit type to hold only four 8bit elements. In clang we don't see this intermediate types but they are still there implicitly.